kernel/src/process.rs

```diff
@@ -1857,6 +1756,144 @@ impl<C: 'static + Chip> Process<'a, C> {
         Ok((Some(process), memory_padding_size + memory_size))
     }
 
+    /// Attempt to restart the process.
+    ///
+    /// This function can be called when the process is in any state and
+    /// attempts to reset all of its state and re-initialize it so that it can
+    /// start running again.
+    ///
+    /// Restarting can fail for two general reasons:
+    ///
+    /// 1. The kernel chooses not to restart the process based on the policy the
+    ///    kernel is using for restarting a specific process. For example, if a
+    ///    process has restarted a number of times in a row the kernel may
+    ///    decide to stop executing it.
+    ///
+    /// 2. Some state can no long be configured for the process. For example,
+    ///    the syscall state for the process fails to initialize.
+    ///
+    /// After `restart()` runs the process will either be queued to run its
+    /// `_start` function, or it will be left in `failure_state`.
+    fn restart(&self, failure_state: State) {
```

Comment:
Since restarting can fail, should the function return some indication of success or failure?

Reply:
I'm not sure. I have no idea what the caller could do with that information, but it might be useful to know what happened and potentially report it back for informative purposes.

But, I think it's better if this PR doesn't change any behavior, just re-organizes code. Actual changes can be part of a different patch.

Reply:
I think if these was a `pub` I'd push back, but given that it's contained, this is fine for now.